reference/distributed/index_map_kernels.cpp

```diff
@@ -0,0 +1,207 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "core/distributed/index_map_kernels.hpp"
+
+
+#include <ginkgo/core/distributed/index_map.hpp>
+
+
+#include "core/base/allocator.hpp"
+#include "core/base/segmented_array.hpp"
+#include "reference/distributed/partition_helpers.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace reference {
+namespace index_map {
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+void build_mapping(
+    std::shared_ptr<const DefaultExecutor> exec,
+    const experimental::distributed::Partition<LocalIndexType, GlobalIndexType>*
+        part,
+    const array<GlobalIndexType>& recv_connections,
+    array<experimental::distributed::comm_index_type>& remote_part_ids,
+    array<LocalIndexType>& remote_local_idxs,
+    array<GlobalIndexType>& remote_global_idxs, array<int64>& remote_sizes)
+{
+    using experimental::distributed::comm_index_type;
+    auto part_ids = part->get_part_ids();
+
+    vector<GlobalIndexType> unique_indices(recv_connections.get_size(), {exec});
+    std::copy_n(recv_connections.get_const_data(), recv_connections.get_size(),
+                unique_indices.begin());
+
+    auto find_part = [&](GlobalIndexType idx) {
+        auto range_id = find_range(idx, part, 0);
+        return part_ids[range_id];
+    };
+
+    // sort by (part-id, global-id)
+    std::sort(unique_indices.begin(), unique_indices.end(),
+              [&](const auto a, const auto b) {
+                  auto part_a = find_part(a);
+                  auto part_b = find_part(b);
+                  return std::tie(part_a, a) < std::tie(part_b, b);
+              });
+
+    // make unique by (part-id, global-id)
+    auto unique_indices_end =
+        std::unique(unique_indices.begin(), unique_indices.end(),
+                    [&](const auto a, const auto b) {
+                        auto part_a = find_part(a);
+                        auto part_b = find_part(b);
+                        return std::tie(part_a, a) == std::tie(part_b, b);
+                    });
```

Comment:
it's annoying that there is no suitable implementation to handle both < and ==, then we could get rid of this almost duplication

Reply:
starship operator <=> pew pew